src/lib/components/FornaContainer.react.js

```diff
@@ -0,0 +1,206 @@
+                this._fornaContainer.addRNA(seq.structure, unpackedOptions);
+            });
+        }
+    }
+
+    shouldComponentUpdate(nextProps) {
+        const {sequences, nodeFillColor, colorScheme} = this.props;
+
+        // update the component if the actual sequences have changed
+        if (
+            (sequences && !nextProps.sequences) ||
+            (!sequences && nextProps.sequences) ||
+            sequences.length !== nextProps.sequences.length ||
+            sequences.some(
+                (seq, i) =>
+                    sequences[i].sequence !== nextProps.sequences[i].sequence ||
+                    sequences[i].structure !==
+                        nextProps.sequences[i].structure ||
+                    (sequences[i].options && !nextProps.sequences[i].options) ||
+                    (!sequences[i].options && nextProps.sequences[i].options) ||
+                    (sequences[i].options &&
+                        nextProps.sequences[i].options &&
+                        Object.keys(seq.options).some(
+                            optionName =>
+                                (sequences[i].options[optionName] &&
+                                    !nextProps.sequences[i].options[
+                                        optionName
+                                    ]) ||
+                                (!sequences[i].options[optionName] &&
+                                    nextProps.sequences[i].options[
+                                        optionName
+                                    ]) ||
+                                sequences[i].options[optionName] !==
+                                    nextProps.sequences[i].options[optionName]
+                        ))
+            )
+        ) {
+            return true;
+        }
+
+        // only change node fill color/color scheme if these props have changed
+        if (nodeFillColor !== nextProps.nodeFillColor) {
+            this._fornaContainer.setOutlineColor(nextProps.nodeFillColor);
+        }
+        if (colorScheme !== nextProps.colorScheme) {
+            this._fornaContainer.changeColorScheme(nextProps.colorScheme);
+        }
+
+        return false;
+    }
+
+    render() {
+        this.renderNewSequences();
+        return <div id={this.props.id} />;
+    }
+}
+
+FornaContainer.propTypes = {
+    /**
+     * The ID of this component, used to identify dash components in
+     * callbacks. The ID needs to be unique across all of the
+     * components in an app.
+     */
+    id: PropTypes.string.isRequired,
+
+    /**
+     * The height (in px) of the container in which the molecules will
+     * be displayed.
+     */
+    height: PropTypes.number,
+
+    /**
+     * The width (in px) of the container in which the molecules will
+     * be displayed.
+     */
+    width: PropTypes.number,
+
+    /**
+     * The molecules that will be displayed.
+     */
+    sequences: PropTypes.arrayOf(
+        PropTypes.shape({
```

Comment:
`PropTypes.exact` instead of `shape`?

Reply:
The user doesn't have to specify a value for `options`, so `PropTypes.shape` applies here (please correct me if I'm wrong)!

Reply:
yeah the naming is tricky but I think `exact` is what you want (in fact we almost never want `shape`). The only difference, in my understanding, is whether extra, non-enumerated keys are allowed - `shape` would not give you a warning/error if you include other things in the object, but `exact` will. `.isRequired` is still honored either way - so you can send an object to `exact` that's missing some optional keys.

From https://www.npmjs.com/package/prop-types:
```js
// An object taking on a particular shape
  optionalObjectWithShape: PropTypes.shape({
    optionalProperty: PropTypes.string,
    requiredProperty: PropTypes.number.isRequired
  }),
 
  // An object with warnings on extra properties
  optionalObjectWithStrictShape: PropTypes.exact({
    optionalProperty: PropTypes.string,
    requiredProperty: PropTypes.number.isRequired
  }),
```